### ISSUE TITLE
VSTS Capitalizes all build variables

### ISSUE THREAD

**Issue by dansiegel** [dansiegel/Mobile.BuildTools#47]
# Description

When adding environment variables in VSTS a variable added to the Build Definition like `HelloWorld` will be added to the environment as `HELLOWORLD` making it harder to discover and impossible to ensure that the correct variable name is used when generating the json file and thus the Secrets Class.

## Reproduction Steps

Create a build definition for a UWP project in App Center or any build definition in VSTS with a Windows Client (I believe this may also be occurring on the Mac agent with VSTS). Add variables to the build definition. All variables added to the environment are capitalized.

## Expected Behavior

If a variable is expected to be in the json as HelloWorld, we need a way of compensating for environments that prevent us from trusting the variable name is in the format we expect.

## Actual Behavior

Environment variables are either undiscovered or discovered incorrectly resulting in capitalization errors that cause the build to fail.

## Environment

- OS: Windows
- IDE: n/a
- BuildTools Version: all
- IDE Version: n/a
- CI Environment: App Center - Windows / VSTS

**Comment by rspaulino** [comment]
Hi @dansiegel, I'm not sure if this is the same problem but I'm using AppCenter and I have a Xamarin.Forms targeting iOS, Android and UWP the first too work but the UWP have issues generating the Secrets class, Do you have any workaround or idea on how to fix this? Thanks in advance and I love your projects.

**Broken**
BuildHostSecrets:
  Output Path: D:\a\1\s\src\MobileTemplate.Core\secrets.json
      No Build Host Secrets Found...
Secrets:
  No Secrets Json File was found.
CoreCompile:
  C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn\csc.exe /noconfig /unsafe- /checked- /nowarn:1701,1702,1701,1702 /nostdlib+ /errorreport:prompt /warn:4 /define:TRACE;DEBUG;NETSTANDARD;NETSTANDARD2_0 /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.appcenter.analytics\1.9.0\lib\netstandard1.0\Microsoft.AppCenter.Analytics.dll /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.appcenter.crashes\1.9.0\lib\netstandard1.0\Microsoft.AppCenter.Crashes.dll /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.appcenter\1.9.0\lib\netstandard1.0\Microsoft.AppCenter.dll /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.codeanalysis.csharp\2.8.2\lib\netstandard1.3\Microsoft.CodeAnalysis.CSharp.dll /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.codeanalysis.csharp.workspaces\2.8.2\lib\netstandard1.3\Microsoft.CodeAnalysis.CSharp.Workspaces.dll /reference:C:\Users\VssAdministrator\.nuget\packages\microsoft.codeanalysis....
  Using shared compilation with compiler from directory: C:\Program Files (x86)\Microsoft Visual Studio\2017\Enterprise\MSBuild\15.0\Bin\Roslyn
##[error]src\MobileTemplate.Core\AppStart.cs(8,27): Error CS0234: The type or namespace name 'Helpers' does not exist in the namespace 'MobileTemplate.Core' (are you missing an assembly reference?)
AppStart.cs(8,27): error CS0234: The type or namespace name 'Helpers' does not exist in the namespace 'MobileTemplate.Core' (are you missing an assembly reference?) [D:\a\1\s\src\MobileTemplate.Core\MobileTemplate.Core.csproj]
Done Building Project "D:\a\1\s\src\MobileTemplate.Core\MobileTemplate.Core.csproj" (default targets) -- FAILED.

**WORKING**
TemplateManifest:
Skipping target "TemplateManifest" because it has no outputs.
BuildHostSecrets:
  Output Path: /Users/vsts/agent/2.140.2/work/1/s/src/MobileTemplate.Core/secrets.json
  Generating secrets.json for Secret_
  {
    "AppCenter_Android_Secret": "*****",
    "AppCenter_UWP_Secret": "*****",
    "AppCenter_iOS_Secret": "*****"
  }
Secrets:
  ProjectBasePath: /Users/vsts/agent/2.140.2/work/1/s/src/MobileTemplate.Core
  SecretsClassName: Secrets
  SecretsJsonFilePath: /Users/vsts/agent/2.140.2/work/1/s/src/MobileTemplate.Core/secrets.json
  RootNamespace: MobileTemplate.Core
  OutputPath: /Users/vsts/agent/2.140.2/work/1/s/src/MobileTemplate.Core/Helpers
  // ------------------------------------------------------------------------------
  //  <autogenerated>
  //      This code was generated by Mobile.BuildTools. For more information or to
  //      file an issue please see https://github.com/dansiegel/Mobile.BuildTools
  //
  //      Changes to this file may cause incorrect behavior and will be lost when 
  //      the code is regenerated. 
  //
  //      NOTE: This file should be excluded from source control.
  //  </autogenerated>
  // ------------------------------------------------------------------------------
  
  namespace MobileTemplate.Core.Helpers
  {
      internal static class Secrets
      {
          internal const string AppCenter_Android_Secret = "*****";
  
          internal const string AppCenter_UWP_Secret = "*****";
  
          internal const string AppCenter_iOS_Secret = "*****";
      }
  }
  
  Writing Secrets Class to: 'obj/iPhoneSimulator/Debug/netstandard2.0/Secrets.cs'

**Comment by rspaulino** [comment]
@dansiegel VSTS have the same problem as well
<img width="1116" alt="screen shot 2018-10-08 at 7 18 36 am" src="https://user-images.githubusercontent.com/378671/46606320-7af6cf80-caca-11e8-8b02-de0c13d0b945.png">

**Comment by xyashapyx** [comment]
I'm having the same issues, despite that fact that on app center all works great.

**Comment by rspaulino** [comment]
@xyashapyx Do you have a UWP project using Mobile.BuildTools building in AppCenter? because I can't build my UWP project on AppCenter or VSTS but the iOS and Android build fine. 

My project is a Xamarin.Forms.

**Comment by xyashapyx** [comment]
@rspaulino Nope, I'm having it for IOS and Android too. Have you configured something additionally for building all except UWP in VSTS?

**Comment by rspaulino** [comment]
@xyashapyx Oh ok I got it. I'm using AppCenter to build iOS/Android/UWP but since my UWP wasn't building I give it a try with VSTS and I got it working.

I will try to compile my iOS/Android project and let you know.

any reason why you don't use AppCenter for your iOS/Android project?

**Comment by xyashapyx** [comment]
@rspaulino I need to add additional code security checks to the building process and it doable with VSTS plug-ins usually. As far as I know App Center doesn't support VSTS plugins.

**Comment by dansiegel** [comment]
@xyashapyx if you need to use the marketplace plugins, you almost certainly should use Azure Pipelines instead of App Center. Though you should remember that you have full access to install anything you need to on the build host with App Center, so if you can script it, you can do it.

Azure Pipelines (formerly VSTS) actually has a number of problems when creating builds with Visual designer. One of the problems in Azure Pipelines is that when you use the naming convention `Secret_VarName`, it treats the variable as if it is actually a secret hiding it from the environment. Interestingly, this isn't 100% accurate, as sometimes it works as expected though it generally just capitalizes the variable if it adds it at all. Another problem is that if you mark a variable as a protected variable (preventing someone from seeing the value), it is also hidden from the environment making it impossible to use. As you will see in the snippet below however, with a YAML build definition these problems go away as you can specify whatever you want, and inject the environment variables you need at the build step.

```yaml
- task: XamariniOS@2
  displayName: 'Build Xamarin.iOS solution **/*.iOS.csproj'
  inputs:
    solutionFile: '**/*.iOS.csproj'
    configuration: $(BuildConfiguration)
  env:
    Manifest_AADClientId: $(ClientId)
    Secret_ClientId: $(ClientId)
    Secret_AppCenterSecret: $(AppCenterSecret)
    Secret_Tenant: $(Tenant)
    Secret_AppServiceEndpoint: $(AppServiceEndpoint)
```

@rspaulino I actually do use App Center for most of my builds. From an economic standpoint though it is sometimes more desirable to use Azure Pipelines as you may only want to pay for one or two concurrent builds with Azure Pipelines, instead of paying for one in Azure Pipelines and another in App Center. I'm sure at some point they will ultimately merge these build pipelines since Azure Pipelines is actually what App Center is using behind the scenes, and the App Center team actually maintains the MacOS build hosts for Azure Pipelines.

As for a work around for UWP what I found was that I had to do a post-clone script. There were actually a couple of problems I ran into while building UWP with App Center:

1. Is the Capitalization issue as noted by this issue
1. When using an authenticated Package Feed such as an Azure Artifacts Package Feed, the NuGet restore task would not pick up the NuGet.config and thus would fail to restore packages for the project.
1. Also when using an authenticated Package Feed, the docs say to place a NuGet.config in the root directory. The problem with this is that you cannot/should not have the auth token in the config in Source Control for security reasons. This means that you are either risking accidently checking in a file that you have modified with your auth token, or that you have a templated file that Visual Studio/NuGet will automatically pick up during local development, and will fail to properly use since it's templated. Yet another issue I found was that the App Center team didn't coordinate between UWP and iOS/Android as to how to template the NuGet.config for automatic Regex replacement.

The work around I had for these issues was as follows:

```ps1
if($env:PackageFeedPat) {
    # NuGet config named template.config to prevent Visual Studio from picking it up for local development
    (Get-Content .\build\template.config) `
      -replace '%PackageFeedPat%', "$env:PackageFeedPat" |
    Out-File NuGet.config
    Write-Host "Generated NuGet.config"
}
else {
    Write-Warning "There is no environment variable defined for the Package Feed"
}

Write-Host "Downloading latest nuget.exe"
Invoke-WebRequest -Uri https://dist.nuget.org/win-x86-commandline/latest/nuget.exe -OutFile .\nuget.exe

Write-Host "Performing NuGet Restore"
.\nuget.exe restore -ConfigFile .\NuGet.config

# Hack... Windows client capitalizes all environment variables wrecking the Mobile.BuildTools
$secrets = @{
    AppCenterSecret = "$env:AppCenterSecret"
    ApplicationID = "$env:ApplicationID"
    TenantID = "$env:TenantID"
    ReturnUri = "$env:ReturnUri"
    ApiResourceUri = "$env:ApiResourceUri"
    ApiResourceId = "$env:ApiResourceId"
    TenantUrl = "$env:TenantUrl"
    SyncfusionLicense = "$env:SyncfusionLicense"
}

$secrets | ConvertTo-Json | Out-File .\src\MyProject\secrets.json
```

**Comment by rspaulino** [comment]
@xyashapyx Ok I got it working on VSTS here is the hack that work for me, please let me know if this work for you too.
In the pipeline I set the following variable depending on the platform:
_SdkShortFrameworkIdentifier = xamarin.ios
iOSSecret_APPCENTER_IOS_SECRET = *****
iOSSecret_APPCENTER_ANDROID_SECRET  = *****

<img width="863" alt="screen shot 2018-10-09 at 8 48 54 am" src="https://user-images.githubusercontent.com/378671/46670438-33db0e00-cba0-11e8-9308-5bd42e01def5.png">
<img width="1052" alt="screen shot 2018-10-09 at 8 48 08 am" src="https://user-images.githubusercontent.com/378671/46670439-33db0e00-cba0-11e8-9a55-ff0d319aa3ff.png">

**Comment by rspaulino** [comment]
@dansiegel Thanks

I end up creating a pre build script to hack it but your solution is way more elegant than mine.

I appreciate your help.

It's just really annoying MS have different experience for a system that is build on top of another system and the lack of UWP support comparing to iOS/Android it just make me want to not even bother with UWP.

**Comment by stevenengland** [comment]
> Azure Pipelines (formerly VSTS) actually has a number of problems when creating builds with Visual designer. One of the problems in Azure Pipelines is that when you use the naming convention `Secret_VarName`, it treats the variable as if it is actually a secret hiding it from the environment. Interestingly, this isn't 100% accurate, as sometimes it works as expected though it generally just capitalizes the variable if it adds it at all. Another problem is that if you mark a variable as a protected variable (preventing someone from seeing the value), it is also hidden from the environment making it impossible to use. As you will see in the snippet below however, with a YAML build definition these problems go away as you can specify whatever you want, and inject the environment variables you need at the build step.
> 
> ```yaml
> - task: XamariniOS@2
>   displayName: 'Build Xamarin.iOS solution **/*.iOS.csproj'
>   inputs:
>     solutionFile: '**/*.iOS.csproj'
>     configuration: $(BuildConfiguration)
>   env:
>     Manifest_AADClientId: $(ClientId)
>     Secret_ClientId: $(ClientId)
>     Secret_AppCenterSecret: $(AppCenterSecret)
>     Secret_Tenant: $(Tenant)
>     Secret_AppServiceEndpoint: $(AppServiceEndpoint)
> ```

For me the only way to get things working using a yaml file was to override the BuildHostSecretPrefix (in the .csproj file)
```
<PropertyGroup>
    <BuildHostSecretPrefix>SAFE_</BuildHostSecretPrefix>
 </PropertyGroup>
```
Then the variables were recognized:
```
- task: XamarinAndroid@1
  displayName: 'Build Xamarin.Android project **/*Droid*.csproj'
  inputs:
    projectFile: '**/*Droid*.csproj'
    outputDirectory: '$(build.binariesdirectory)/$(BuildConfiguration)'
    configuration: '$(BuildConfiguration)'
    msbuildArguments: '/verbosity:detailed'
  env:
    SAFE_APPCENTER_IOS_SECRET: $(AppCenter_iOS_Secret)
    SAFE_APPCENTER_ANDROID_SECRET: $(AppCenter_Android_Secret)
    SAFE_APPCENTER_WINDOWS_SECRET: $(AppCenter_Windows_Secret)
    SolutionDir: $(Build.Repository.LocalPath)
```

**Comment by soydachi** [comment]
@stevenengland thanks for you comment, this solutions work for me in Azure Pipeline, so no need to create an YAML pipeline. In the visual design mode it's works too.

**Comment by dansiegel** [comment]
After thinking a lot on this the only reliable way to handle this is to require a property be added with the keys. Since we would only be exposing keys which are already exposed in code as properties of the Secrets class this would still be a secure way of injecting build time configurations and keys.

```xml
<PropertyGroup>
  <BuildHostKeys>Foo;Bar</BuildHostKeys>
</PropertyGroup>
```

This would then search for each key regardless of capitalization and ensure that the property is added as Foo or Bar instead of FOO and BAR

**Comment by dansiegel** [comment]
will be fixed by #72
